main.go

```diff
@@ -160,12 +155,38 @@ func main() {
 		}
 	}
 	// Fetch the configuration for maximum number of API keys allowed per user.
-	maxAPIKeys, err := strconv.Atoi(os.Getenv(envMaxNumAPIKeysPerUser))
+	maxAPIKeysStr := os.Getenv(envMaxNumAPIKeysPerUser)
+	if maxAPIKeysStr != "" {
+		maxAPIKeys, err := strconv.Atoi(maxAPIKeysStr)
+		if err != nil {
+			log.Printf("Warning: Failed to parse %s env var. Error: %s", envMaxNumAPIKeysPerUser, err.Error())
+		}
+		if maxAPIKeys > 0 {
+			database.MaxNumAPIKeysPerUser = maxAPIKeys
+		}
+	}
+
+	return emailURI, nil
+}
+
+func main() {
+	// Initialise the global context and logger. These will be used throughout
+	// the service. Once the context is closed, all background threads will
+	// wind themselves down.
+	ctx := context.Background()
+	logger := logrus.New()
+	logger.SetLevel(logLevel())
+
+	// Load the environment variables from the .env file.
+	_ = godotenv.Load()
+	dbCreds, err := loadDBCredentials()
 	if err != nil {
-		log.Printf("Warning: Failed to parse %s env var. Error: %s", envMaxNumAPIKeysPerUser, err.Error())
+		log.Fatal(errors.AddContext(err, "failed to fetch DB credentials"))
 	}
-	if maxAPIKeys > 0 {
-		database.MaxNumAPIKeysPerUser = maxAPIKeys
+
+	emailURI, err := parseEnvironmentVariables(logger)
```

Comment:
I think it's a great idea to improve this, but in this case I think it's a bit quirky in the sense that sometimes it returns the env variable and other times it will go and overwrite exposed variables on other packages. It's quite magical right, you call it and you don't know what's going to happen at all.

I found that some of the variables don't mention in the docstring they are (possibly) overwritten by a value on the environment, I think `jwt.TTL` is an example of where it's missing. I dislike that pattern personally but that's just personal preference I guess, it's quite annoying to debug IMO because if you would look for the TTL, you'd find a certain value and not know whether it got overwritten, in this case you wouldn't even know it gets overwritten, but that's just because of the oversight of mentioning it in the docstring, which is easily overlooked.

Reply:
That's a valid concern, on both counts.

On the side of variables that do not state which env vars control them the solution is pretty obvious - I'll just update their docstrings.

On the side of variables being directly set on packages instead of being returned - I'm not sure how I want to attack that.
* Having the parsing method return a bunch of variables, so we can set them in `main` doesn't sound appealing. 
* Adding a struct that defines all env vars and using that as return value doesn't improve the situation much. 
* Moving all these values to a separate package might be a way to go but it will move the variables themselves away from the code that uses them. The initialisation flow will be cleaner but the actual usage will be less so.

I am open to suggestions here and I don't have a preference.

Reply:
I agree with PJ. Modifying global vars outside of the scope of a function is a nightmare to maintain. 
I wouldn't expect a parsing method to actually set anything that's not passed in by reference or returned. Instead I'd want it to return a struct with some configuration values which I then pass on to whatever needs them.

Reply:
OK, in that case I'll define a configuration type and I'll return it from the parsing func. Then I'll set the global vars from `main` as before. This should achieve both the visibility we had before and the testability we wanted to get from this PR. It will also simplify `main` quite a bit, something I wanted to get.